src/site/content/en/blog/asyncify/index.md

```diff
@@ -0,0 +1,516 @@
+
+Luckily, the Asyncify transform itself is completely toolchain-agnostic. It can transform arbitrary
+WebAssembly files, no matter which compiler it's produced by. The transform is provided separately
+as part of the `wasm-opt` optimiser from the [Binaryen
+toolchain](https://github.com/WebAssembly/binaryen) and can be invoked like this:
+
+```shell
+wasm-opt -O2 --asyncify \
+      --pass-arg=asyncify-imports@env.get_answer \
+      [...]
+```
+
+Pass `--asyncify` to enable the transform, and then use `--pass-arg=…` to provide a comma-separated
+list of asynchronous functions, where the program state should be suspended and later resumed.
+
+All that's left is to provide supporting runtime code that will actually do that—suspend and resume
+our code. Again, in the C / C++ case this would be included by Emscripten, but now we need custom
+JavaScript glue code that would handle arbitrary WebAssembly files. We've created a library just for
+that.
+
+You can find it on Github at
+[https://github.com/GoogleChromeLabs/asyncify](https://github.com/GoogleChromeLabs/asyncify) or npm
+under the name [`asyncify-wasm](https://www.npmjs.com/package/asyncify-wasm)`.
+
+It simulates a standard [WebAssembly instantiation
+API](https://developer.mozilla.org/en-US/docs/WebAssembly), but under its own namespace. The only
+difference is that, under a regular WebAssembly API you can only provide synchronous functions as
+imports, while under the Asyncify wrapper, you can provide asynchronous imports as well:
+
+```js
+let … = await Asyncify.instantiateStreaming(fetch('app.wasm'), {
+    env: {
+        async get_answer() {
+            let response = await fetch("answer.txt");
+            let text = await response.text();
+            return Number(text);
+        }
+    }
+});
+…
+await instance.exports.main();
+```
+
+Once you try to call such asynchronous function - like `get_answer()` in the example above - from
+the WebAssembly side, the library will detect the returned `Promise`, suspend and save the state of
+the WebAssembly application, subscribe to the promise completion, and later, once it's resolved,
+seamlessly restore the call stack and state and continue execution as if nothing has happened.
+
+Since any function in the module might make an asynchronous call, all the exports become potentially
+asynchronous, too, so they get wrapped as well. You might have noticed in the example above that you
+need to `await` the result of `instance.exports.main()` to know when the execution is truly
+finished.
+
+### How does this all work under the hood?
+
+When Asyncify detects a call to one of the `ASYNCIFY_IMPORTS` functions, it starts an asynchronous
+operation, saves the entire state of the application, including the call stack and any temporary
+locals, and later, when that operation is finished, restores all the memory and call stack and and
+resumes from the exact same place and with the exact same state as if the program has never stopped.
```

Comment:
```suggestion
resumes from the exact place and with the exact state as if the program has never stopped.
```

Reply:
Now it's unclear which exact place / state it's referring to - something we pass as an argument or ..? IMO "same" is important to make it clear it's still the same place/state as before.

Reply:
I was trying to remove some wordiness. What about removing 'exact' instead?

Reply:
That works.

```suggestion
resumes from the same place and with the same state as if the program has never stopped.
```